corrects missleading information in "creating a scene" example

GGAlanSmithee opened the pull request:
The documentation states that invoking setSize with a lower value then the viewport will cause the scene to be rendered at a lower resolution, which is not true.

The correct way to do this is to use setPixelRatio. See [this](http://stackoverflow.com/questions/31407778/display-scene-at-lower-resolution-in-three-js) SO question.

I have corrected the documentation to reflect this.

GGAlanSmithee committed:
corrects missleading information in "creating a scene" example

mrdoob commented on docs/manual/introduction/Creating-a-scene.html:
```diff
@@ -63,7 +63,7 @@
 		<div>Next up is the renderer. This is where the magic happens. In addition to the WebGLRenderer we use here, Three.js comes with a few others, often used as fallbacks for users with older browsers or for those who don't have WebGL support for some reason.</div>
 
-		<div>In addition to creating the renderer instance, we also need to set the size at which we want it to render our app. It's a good idea to use the width and height of the area we want to fill with our game - in this case, the width and height of the browser window. For performance intensive games, you can also give <strong>setSize</strong> smaller values, like <strong>window.innerWidth/2</stro ...
+		<div>In addition to creating the renderer instance, we also need to set the size at which we want it to render our app. It's a good idea to use the width and height of the area we want to fill with our game - in this case, the width and height of the browser window. For performance intensive games, you can also give <strong>setSize</strong> smaller values, like <strong>window.innerWidth/2</stro ...
```
Hmm, why are you mentioning a "game"?

GGAlanSmithee commented:
If you compare the old and new sentence, "game" is being mentioned twice already. I was just following the prior writers style.

I am more than happy to change all mentions of a "game" to "app" if you think that is more appropriate.

mrdoob commented:
Ah! Sorry, I didn't realise it was mentioned already.

WestLangley commented:
> I am more than happy to change all mentions of a "game" to "app"

I think that is preferable.

mrdoob commented:
I think so too.

dubejf commented:
The reason why your display area shrinks when you do `renderer.setSize( window.innerWidth/2, window.innerHeight/2 )` is that you have not set your css style for the canvas, `{ width: 100%; height: 100% }`, as shown in the `getting started` example.

Canvas need a css style, otherwise they will be displayed using the same size as the drawing buffer.

`renderer.setSize` accepts a third argument, `updateStyle`, which defaults to `true`. So, when you call, `renderer.setSize( window.innerWidth/2, window.innerHeight/2 )` you actually update the display size to fill a quarter of the window.

If you want to upscale or downscale:
1) provide a css style to pin the display size,
2) use the 3rd argument when setting the size of your drawing buffer: `renderer.setSize( window.innerWidth/2, window.innerHeight/2, false )`

I think that `setPixelRatio` should be used with `window.devicePixelRatio` (the device dpi or browser zoom factor) in the tutorial. Or, at least `renderer.setPixelRatio( 0.5 * window.devicePixelRatio )` instead of `renderer.setPixelRatio( 0.5 )`.

But I still think that setting the css size explicitly should be the recommended method!

More info: http://webglfundamentals.org/webgl/lessons/webgl-resizing-the-canvas.html

GGAlanSmithee commented:
Interesting!

I think there is some value to setting resolution via the API rather than by explicitly doing it with CSS since it allows for the implementation to change without forcing users to refactor. If the recommended way to do so is by giving the canvas a size and than pass false as updateStyle when calling setSize, then thats great.

Regarding setPixelRatio. If the only usecase for this is to set device pixel ratio, why does this function exists at all? Shouldnt setDevicePixelRatio be enough?

I still think this PR has som value since the example is often the first contact a user will have with three. It needs to not be confusing. I suggest that I do the following changes to this PR:

* revert to the original text
* replace all mentions of "game" with "app"
* clarify that you need to set a size to the canvas and pass false to updateStyle in order for the expected behaviour to occur

Thoughts?

mrdoob commented:
> I think that `setPixelRatio` should be used with `window.devicePixelRatio` (the device dpi or browser zoom factor) in the tutorial. Or, at least `renderer.setPixelRatio( 0.5 * window.devicePixelRatio )` instead of `renderer.setPixelRatio( 0.5 )`.

I like this.

dubejf commented:
@GGAlanSmithee There are certainly ways to improve the example and the API.

Maybe this would improve the API:
- Add `.autoResize` to the renderer (defaults to `true`).
- Remove third parameter from setSize. Style will be update based on `.autoResize`.
- Extract `setClientSize( width, height )` method to set the style size of the canvas and add it to the renderer.

Usage:
```
// Normal
renderer.setSize( window.innerWidth, window.innerHeight );
renderer.setPixelRatio( window.devicePixelRatio );
```
```
// With upscaling
renderer.autoResize = false;
renderer.setClientSize( window.innerWidth, window.innerHeight );
renderer.setSize( 512, 512 );
renderer.setPixelRatio( window.devicePixelRatio );
```

Maybe `autoResize` is misleading - people might that that it involves a resize handler...

There are a lot of good suggestions in the issue tracker (look for "devicePixelRatio") - I confess that I have not paid close attention to them.

GGAlanSmithee commented:
Ah. Nvm my comment about devicePixelRatio, I thought it was a part of threejs. What you said makes more sense now :)

In regards to your comment about using it with devicePixelRatio, should setPixelRatios implementation change to factor it in? I guess not since that would be anything but backwards compatible.

GGAlanSmithee commented:
Thanks again for the info @dubejf

I will check the issue tracker and see if I can make something useful with your suggestions.

GGAlanSmithee commented:
Reading through the issues and PRs related to devicePixelRatio, I see that the problem is more involved than I though.

From what I've been able to gather, there has been work made (mostly by @greggman) to decouple devicePixelRatio from three.js. The reason being that you want to get well defined and expected behaviour as far as canvas and renderer size is concerned and the trade-off is that the user is responsible for managing the pixel ratio. #4903, #5627, #6027

This seems like a reasonable trade-off to me, since it looks like there has been a lot of issues related to three.js aoutomagically handling canvas scaling. 

About refactoring the API, I am feeling a bit reluctant to do so now, since a) I do not know enough about the usage and impact of the possible changes to make those calls, and b) changing the API (in a non backward compatible way) seems like a bad idea since there has been quite a lot of changes related to this in the last year or so. It does, however, look like #6073 actually suggests something along the lines of what you are talking about @dubejf, adding a <code>autoResize</code> parameter to the renderer.

Within the scope of this PR, I suggest that either:

* the line about changing resolution is removed from the "Creating a scene" example, and that an example dedicated to <code>setPixelRatio</code> is added, since this information is not really relevant in a "Hello, World!" example

or

* the line about changing resolution is updated so that it makes clear that in order for the resolution change to take effect, the canvas needs to have a size and that <code>updateStyle</code> needs to be <code>true</code>

GGAlanSmithee committed:
updates create a scene example

- reverts back to the original text, removing any mention about setPixelRatio
- adds information about how to render an app at lower resolution using updateStyle
- replaces mentions of "game" with "app"

GGAlanSmithee commented:
Another attempt to update the info in "create a scene" example according to the discussion in this PR.
I left out any mention of setPixelRatio since I don't think it is relevant in this context, It could always be included should you deem it important enough - if so let me know.

mrdoob commented:
Thanks!
